mirgecom/profiling.py

```diff
@@ -0,0 +1,229 @@
+    kwargs: dict
+
+
+class PyOpenCLProfilingArrayContext(PyOpenCLArrayContext):
+    """An array context that profiles kernel executions.
+
+    .. automethod:: tabulate_profiling_data
+    .. automethod:: call_loopy
+
+    Inherits from :class:`meshmode.array_context.PyOpenCLArrayContext`.
+    """
+
+    def __init__(self, queue, allocator=None) -> None:
+        super().__init__(queue, allocator)
+
+        if not queue.properties & cl.command_queue_properties.PROFILING_ENABLE:
+            raise RuntimeError("Profiling was not enabled in the command queue. "
+                 "Please create the queue with "
+                 "cl.command_queue_properties.PROFILING_ENABLE.")
+
+        self.profile_events = []
+        self.profile_results = {}
+        self.kernel_stats = {}
+
+    def _finish_profile_events(self) -> None:
+        # First, wait for completion of all events
+        if self.profile_events:
+            cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
+
+        # Then, collect all events and store them
+        for t in self.profile_events:
+            program = t.program
+            r = self._get_kernel_stats(program, t.kwargs)
+            time = t.cl_event.profile.end - t.cl_event.profile.start
+
+            new = ProfileResult(time, r.flops, r.bytes_accessed, r.footprint_bytes)
+
+            self.profile_results.setdefault(program, []).append(new)
+
+        self.profile_events = []
+
+    def tabulate_profiling_data(self) -> pytools.Table:
+        """Return a :class:`pytools.Table` with the profiling results."""
+        self._finish_profile_events()
+
+        tbl = pytools.Table()
+
+        tbl.add_row(["Function", "Calls",
+            "Time_min [s]", "Time_avg [s]", "Time_max [s]",
+            "GFlops/s_min", "GFlops/s_avg", "GFlops/s_max",
+            "BWAcc_min [GByte/s]", "BWAcc_mean [GByte/s]", "BWAcc_max [GByte/s]",
+            "BWFoot_min [GByte/s]", "BWFoot_mean [GByte/s]", "BWFoot_max [GByte/s]"])
+
+        # Precision of results
+        g = ".4g"
+
+        from statistics import mean
+
+        for key, value in self.profile_results.items():
+            num_values = len(value)
+
+            times = [v.time / 1e9 for v in value]
+            flops = [v.flops / 1e9 for v in value]
+
+            flops_per_sec = [f / t for f, t in zip(flops, times)]
+
+            bytes_accessed = [v.bytes_accessed / 1e9 for v in value]
+            footprint_bytes = [v.footprint_bytes / 1e9 for v in value
+                if v.footprint_bytes is not None]
+
+            bandwidth_access = [b / t for b, t in zip(bytes_accessed, times)]
+
+            if len(footprint_bytes) > 0:
+                bandwidth_footprint = [b / t for b, t in zip(footprint_bytes, times)]
+                bandwidth_footprint_val = {"min": f"{min(bandwidth_footprint):{g}}",
+                    "mean": f"{mean(bandwidth_footprint):{g}}",
+                    "max": f"{max(bandwidth_footprint):{g}}"}
```

Comment:
Numpy's masked arrays do this with less fuss:
```py
>>> ary = np.ma.masked_equal([12, 13, None], None)
>>> ary
masked_array(data=[12, 13, --],
             mask=[False, False,  True],
       fill_value='?',
            dtype=object)
>>> np.mean(ary)
12.5
>>> ary = np.ma.masked_equal([None,None, None], None)
>>> f"{np.mean(ary):.4g}"
'--'
>>> 
```

Reply:
Fixed. (np.min/np.max are unfortunately a bit harder than np.mean)